jumanji/environments/logic/graph_coloring/env.py

```diff
@@ -0,0 +1,264 @@
+        - colors: jax array (int32) of shape (num_nodes,):
+            color assigned to each node, -1 if not assigned
+        - adj_matrix: jax array (bool) of shape (num_nodes, num_nodes):
+            adjacency matrix of the graph
+    """
+
+    def __init__(
+        self,
+        generator: Optional[Generator] = None,
+    ):
+        """Instantiate a `GraphColoring` environment.
+
+        Args:
+            generator: callable to instantiate environment instances.
+                Defaults to `RandomGenerator` with `num_nodes = 100`,
+                and `percent_connected = 0.8` parameters.
+        """
+        self.generator = generator or RandomGenerator(
+            num_nodes=100,
+            percent_connected=0.8,
+        )
+        num_nodes, percent_connected = self.generator.specs()
+        # Create viewer used for rendering
+        self._env_viewer = GraphColoringViewer(
+            num_nodes=num_nodes, name="Graph Coloring"
+        )
+
+    def __repr__(self) -> str:
+        """Returns: str: the string representation of the environment."""
+        num_nodes, percent_connected = self.generator.specs()
+        return (
+            f"Graph coloring(number of nodes={num_nodes}, "
+            f"percent connected={percent_connected * 100}% "
+        )
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """Resets the environment to its initial state
+
+        Returns:
+            the initial state and timestep.
+        """
+        num_nodes, percent_connected = self.generator.specs()
+        colors = jnp.full(num_nodes, -1, dtype=jnp.int32)
+        key, subkey = jax.random.split(key)
+        adj_matrix = self.generator(subkey)
+
+        action_mask = jnp.ones(num_nodes, dtype=bool)
+        state = State(
+            adj_matrix=adj_matrix,
+            colors=colors,
+            current_node_index=jnp.array(0),
+            key=key,
+        )
+        obs = Observation(
+            adj_matrix=adj_matrix,
+            colors=colors,
+            action_mask=action_mask,
+            current_node_index=jnp.array(0),
+        )
+        timestep = restart(observation=obs)
+
+        return state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        num_nodes, _ = self.generator.specs()
+        # Get the valid actions for the current state.
+        valid_actions = self._get_valid_actions(state)
+        # Check if the chosen action is invalid (not in valid_actions).
+        invalid_action_taken = jnp.logical_not(valid_actions[action])
+
+        # Update the colors array with the chosen action.
+        colors = state.colors.at[state.current_node_index].set(action)
+        colored_nodes = colors >= 0
+
+        # Determine if all nodes have been assigned a color
+        all_nodes_colored = jnp.count_nonzero(colored_nodes) == num_nodes
```

Comment:
The latter assumes that `state.current_node_index` always starts at 0, which may not be the case in the future (if a generator decides otherwise). I agree with the former though!